File: docs/comparison-tests/results-without-mcp/python/test-4-result.py
# Test PY-4: Refactor to Python best practices

from dataclasses import dataclass
from typing import Protocol, Optional
from decimal import Decimal
from enum import Enum
import pytest


# Custom exceptions
class OrderError(Exception):
    """Base exception for order processing errors."""
    pass


class InvalidInputError(OrderError):
    """Raised when order or user input is invalid."""
    pass


class EmptyCartError(OrderError):
    """Raised when cart has no items."""
    pass


class InsufficientStockError(OrderError):
    """Raised when item is out of stock."""
    def __init__(self, item_name: str):
        self.item_name = item_name
        super().__init__(f"Out of stock: {item_name}")


class InsufficientFundsError(OrderError):
    """Raised when user balance is insufficient."""
    pass


class PaymentError(OrderError):
    """Raised when payment processing fails."""
    pass


# Value objects
@dataclass(frozen=True)
class OrderItem:
    id: str
    name: str
    price: Decimal
    quantity: int

    @property
    def subtotal(self) -> Decimal:
        return self.price * self.quantity


@dataclass(frozen=True)
class Order:
    id: str
    items: tuple[OrderItem, ...]

    @property
    def total(self) -> Decimal:
        return sum(item.subtotal for item in self.items)


@dataclass(frozen=True)
class User:
    id: str
    email: str
    balance: Decimal


@dataclass(frozen=True)
class OrderResult:
    success: bool
    order_id: Optional[str] = None
    total: Optional[Decimal] = None
    error: Optional[str] = None


# Protocols (interfaces)
class InventoryProtocol(Protocol):
    def check(self, item_id: str) -> int:
        ...

    def decrease(self, item_id: str, quantity: int) -> None:
        ...


class PaymentGatewayProtocol(Protocol):
    def charge(self, user_id: str, amount: Decimal) -> dict:
        ...


class EmailServiceProtocol(Protocol):
    def send(self, to: str, subject: str, body: str) -> None:
        ...


# Refactored service
class OrderProcessor:
    def __init__(
        self,
        inventory: InventoryProtocol,
        payment_gateway: PaymentGatewayProtocol,
        email_service: EmailServiceProtocol,
    ):
        self._inventory = inventory
        self._payment_gateway = payment_gateway
        self._email_service = email_service

    def process_order(self, order: Optional[Order], user: Optional[User]) -> OrderResult:
        try:
            self._validate_input(order, user)
            self._check_inventory(order)
            self._validate_balance(user, order.total)
            self._process_payment(user, order.total)
            self._update_inventory(order)
            self._send_confirmation(user, order)

            return OrderResult(
                success=True,
                order_id=order.id,
                total=order.total,
            )
        except OrderError as e:
            return OrderResult(success=False, error=str(e))

    def _validate_input(self, order: Optional[Order], user: Optional[User]) -> None:
        if order is None or user is None:
            raise InvalidInputError("Invalid input")
        if not order.items:
            raise EmptyCartError("Empty cart")

    def _check_inventory(self, order: Order) -> None:
        for item in order.items:
            available = self._inventory.check(item.id)
            if available < item.quantity:
                raise InsufficientStockError(item.name)

    def _validate_balance(self, user: User, total: Decimal) -> None:
        if user.balance < total:
            raise InsufficientFundsError("Insufficient funds")

    def _process_payment(self, user: User, total: Decimal) -> None:
        result = self._payment_gateway.charge(user.id, total)
        if not result.get("success"):
            raise PaymentError("Payment failed")

    def _update_inventory(self, order: Order) -> None:
        for item in order.items:
            self._inventory.decrease(item.id, item.quantity)

    def _send_confirmation(self, user: User, order: Order) -> None:
        self._email_service.send(
            user.email,
            "Order confirmed",
            f"Your order #{order.id} is confirmed.",
        )


# Tests
class MockInventory:
    def __init__(self, stock: dict[str, int] = None):
        self._stock = stock or {}
        self.decreases: list[tuple[str, int]] = []

    def check(self, item_id: str) -> int:
        return self._stock.get(item_id, 0)

    def decrease(self, item_id: str, quantity: int) -> None:
        self.decreases.append((item_id, quantity))


class MockPaymentGateway:
    def __init__(self, success: bool = True):
        self._success = success
        self.charges: list[tuple[str, Decimal]] = []

    def charge(self, user_id: str, amount: Decimal) -> dict:
        self.charges.append((user_id, amount))
        return {"success": self._success}


class MockEmailService:
    def __init__(self):
        self.sent: list[tuple[str, str, str]] = []

    def send(self, to: str, subject: str, body: str) -> None:
        self.sent.append((to, subject, body))


class TestOrderProcessor:
    @pytest.fixture
    def inventory(self):
        return MockInventory({"item-1": 100, "item-2": 50})

    @pytest.fixture
    def payment_gateway(self):
        return MockPaymentGateway(success=True)

    @pytest.fixture
    def email_service(self):
        return MockEmailService()

    @pytest.fixture
    def processor(self, inventory, payment_gateway, email_service):
        return OrderProcessor(inventory, payment_gateway, email_service)

    @pytest.fixture
    def sample_order(self):
        return Order(
            id="order-123",
            items=(
                OrderItem(id="item-1", name="Product A", price=Decimal("10.00"), quantity=2),
                OrderItem(id="item-2", name="Product B", price=Decimal("5.00"), quantity=1),
            ),
        )

    @pytest.fixture
    def sample_user(self):
        return User(id="user-1", email="test@example.com", balance=Decimal("100.00"))

    class TestInputValidation:
        def test_returns_error_when_order_is_none(self, processor, sample_user):
            result = processor.process_order(None, sample_user)

            assert result.success is False
            assert result.error == "Invalid input"

        def test_returns_error_when_user_is_none(self, processor, sample_order):
            result = processor.process_order(sample_order, None)

            assert result.success is False
            assert result.error == "Invalid input"

        def test_returns_error_when_cart_is_empty(self, processor, sample_user):
            empty_order = Order(id="order-1", items=())

            result = processor.process_order(empty_order, sample_user)

            assert result.success is False
            assert result.error == "Empty cart"

    class TestInventoryCheck:
        def test_returns_error_when_out_of_stock(self, processor, sample_user):
            order = Order(
                id="order-1",
                items=(OrderItem(id="item-1", name="Widget", price=Decimal("10"), quantity=200),),
            )

            result = processor.process_order(order, sample_user)

            assert result.success is False
            assert "Out of stock: Widget" in result.error

    class TestBalanceValidation:
        def test_returns_error_when_insufficient_funds(self, processor, sample_order):
            poor_user = User(id="user-1", email="test@example.com", balance=Decimal("1.00"))

            result = processor.process_order(sample_order, poor_user)

            assert result.success is False
            assert result.error == "Insufficient funds"

    class TestPayment:
        def test_returns_error_when_payment_fails(self, inventory, email_service, sample_order, sample_user):
            failed_payment = MockPaymentGateway(success=False)
            processor = OrderProcessor(inventory, failed_payment, email_service)

            result = processor.process_order(sample_order, sample_user)

            assert result.success is False
            assert result.error == "Payment failed"

    class TestSuccessfulOrder:
        def test_updates_inventory(self, processor, inventory, sample_order, sample_user):
            processor.process_order(sample_order, sample_user)

            assert ("item-1", 2) in inventory.decreases
            assert ("item-2", 1) in inventory.decreases

        def test_sends_confirmation_email(self, processor, email_service, sample_order, sample_user):
            processor.process_order(sample_order, sample_user)

            assert len(email_service.sent) == 1
            to, subject, body = email_service.sent[0]
            assert to == "test@example.com"
            assert subject == "Order confirmed"
            assert "order-123" in body

        def test_returns_success_result(self, processor, sample_order, sample_user):
            result = processor.process_order(sample_order, sample_user)

            assert result.success is True
            assert result.order_id == "order-123"
            assert result.total == Decimal("25.00")  # 10*2 + 5*1

        def test_charges_correct_amount(self, processor, payment_gateway, sample_order, sample_user):
            processor.process_order(sample_order, sample_user)

            assert len(payment_gateway.charges) == 1
            user_id, amount = payment_gateway.charges[0]
            assert user_id == "user-1"
            assert amount == Decimal("25.00")

File: docs/comparison-tests/results-without-mcp/typescript/test-4-result.ts
// Test TS-4: Refactor processOrder to be more readable and testable

// Original function was monolithic and hard to test

interface OrderItem {
  id: string;
  name: string;
  price: number;
  qty: number;
}

interface Order {
  id: string;
  items: OrderItem[];
}

interface User {
  id: string;
  email: string;
  balance: number;
}

interface Inventory {
  check(itemId: string): number;
  decrease(itemId: string, quantity: number): void;
}

interface PaymentGateway {
  charge(userId: string, amount: number): { success: boolean };
}

interface EmailService {
  send(to: string, subject: string, body: string): void;
}

interface OrderResult {
  success: boolean;
  error?: string;
  orderId?: string;
  total?: number;
}

class OrderValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'OrderValidationError';
  }
}

class OrderProcessor {
  constructor(
    private inventory: Inventory,
    private paymentGateway: PaymentGateway,
    private emailService: EmailService
  ) {}

  processOrder(order: Order | null, user: User | null): OrderResult {
    try {
      this.validateInput(order, user);
      const total = this.calculateTotalAndCheckInventory(order!);
      this.validateUserBalance(user!, total);
      this.processPayment(user!, total);
      this.updateInventory(order!);
      this.sendConfirmationEmail(user!, order!);

      return { success: true, orderId: order!.id, total };
    } catch (error) {
      if (error instanceof OrderValidationError) {
        return { success: false, error: error.message };
      }
      throw error;
    }
  }

  private validateInput(order: Order | null, user: User | null): void {
    if (!order || !user) {
      throw new OrderValidationError('Invalid input');
    }
    if (order.items.length === 0) {
      throw new OrderValidationError('Empty cart');
    }
  }

  private calculateTotalAndCheckInventory(order: Order): number {
    let total = 0;
    for (const item of order.items) {
      const availableStock = this.inventory.check(item.id);
      if (availableStock < item.qty) {
        throw new OrderValidationError(`Out of stock: ${item.name}`);
      }
      total += item.price * item.qty;
    }
    return total;
  }

  private validateUserBalance(user: User, total: number): void {
    if (user.balance < total) {
      throw new OrderValidationError('Insufficient funds');
    }
  }

  private processPayment(user: User, total: number): void {
    const payment = this.paymentGateway.charge(user.id, total);
    if (!payment.success) {
      throw new OrderValidationError('Payment failed');
    }
  }

  private updateInventory(order: Order): void {
    for (const item of order.items) {
      this.inventory.decrease(item.id, item.qty);
    }
  }

  private sendConfirmationEmail(user: User, order: Order): void {
    this.emailService.send(
      user.email,
      'Order confirmed',
      `Your order #${order.id} is confirmed.`
    );
  }
}

// Tests
describe('OrderProcessor', () => {
  let inventory: jest.Mocked<Inventory>;
  let paymentGateway: jest.Mocked<PaymentGateway>;
  let emailService: jest.Mocked<EmailService>;
  let processor: OrderProcessor;

  const createOrder = (items: OrderItem[] = []): Order => ({
    id: 'order-123',
    items,
  });

  const createUser = (balance = 100): User => ({
    id: 'user-1',
    email: 'test@example.com',
    balance,
  });

  const createItem = (overrides: Partial<OrderItem> = {}): OrderItem => ({
    id: 'item-1',
    name: 'Test Item',
    price: 10,
    qty: 1,
    ...overrides,
  });

  beforeEach(() => {
    inventory = {
      check: jest.fn().mockReturnValue(100),
      decrease: jest.fn(),
    };
    paymentGateway = {
      charge: jest.fn().mockReturnValue({ success: true }),
    };
    emailService = {
      send: jest.fn(),
    };
    processor = new OrderProcessor(inventory, paymentGateway, emailService);
  });

  describe('input validation', () => {
    it('should return error when order is null', () => {
      const result = processor.processOrder(null, createUser());
      expect(result).toEqual({ success: false, error: 'Invalid input' });
    });

    it('should return error when user is null', () => {
      const result = processor.processOrder(createOrder([createItem()]), null);
      expect(result).toEqual({ success: false, error: 'Invalid input' });
    });

    it('should return error when cart is empty', () => {
      const result = processor.processOrder(createOrder([]), createUser());
      expect(result).toEqual({ success: false, error: 'Empty cart' });
    });
  });

  describe('inventory check', () => {
    it('should return error when item is out of stock', () => {
      inventory.check.mockReturnValue(0);
      const order = createOrder([createItem({ name: 'Widget', qty: 5 })]);

      const result = processor.processOrder(order, createUser());

      expect(result).toEqual({ success: false, error: 'Out of stock: Widget' });
    });

    it('should check inventory for each item', () => {
      const order = createOrder([
        createItem({ id: 'item-1', qty: 2 }),
        createItem({ id: 'item-2', qty: 3 }),
      ]);

      processor.processOrder(order, createUser());

      expect(inventory.check).toHaveBeenCalledWith('item-1');
      expect(inventory.check).toHaveBeenCalledWith('item-2');
    });
  });

  describe('balance validation', () => {
    it('should return error when user has insufficient funds', () => {
      const order = createOrder([createItem({ price: 50, qty: 3 })]); // Total: 150
      const user = createUser(100);

      const result = processor.processOrder(order, user);

      expect(result).toEqual({ success: false, error: 'Insufficient funds' });
    });
  });

  describe('payment processing', () => {
    it('should return error when payment fails', () => {
      paymentGateway.charge.mockReturnValue({ success: false });
      const order = createOrder([createItem()]);

      const result = processor.processOrder(order, createUser());

      expect(result).toEqual({ success: false, error: 'Payment failed' });
    });

    it('should charge correct amount', () => {
      const order = createOrder([
        createItem({ price: 10, qty: 2 }),
        createItem({ price: 5, qty: 4 }),
      ]);
      const user = createUser(100);

      processor.processOrder(order, user);

      expect(paymentGateway.charge).toHaveBeenCalledWith('user-1', 40);
    });
  });

  describe('successful order', () => {
    it('should update inventory for each item', () => {
      const order = createOrder([
        createItem({ id: 'item-1', qty: 2 }),
        createItem({ id: 'item-2', qty: 3 }),
      ]);

      processor.processOrder(order, createUser());

      expect(inventory.decrease).toHaveBeenCalledWith('item-1', 2);
      expect(inventory.decrease).toHaveBeenCalledWith('item-2', 3);
    });

    it('should send confirmation email', () => {
      const order = createOrder([createItem()]);
      const user = createUser();

      processor.processOrder(order, user);

      expect(emailService.send).toHaveBeenCalledWith(
        'test@example.com',
        'Order confirmed',
        'Your order #order-123 is confirmed.'
      );
    });

    it('should return success result with order details', () => {
      const order = createOrder([createItem({ price: 25, qty: 2 })]);

      const result = processor.processOrder(order, createUser());

      expect(result).toEqual({
        success: true,
        orderId: 'order-123',
        total: 50,
      });
    });
  });
});
